src/coreclr/jit/lowerxarch.cpp

```diff
@@ -6335,38 +6335,51 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                 {
                     if ((intrinsicId >= NI_FMA_MultiplyAdd) && (intrinsicId <= NI_FMA_MultiplySubtractNegatedScalar))
                     {
-                        bool supportsRegOptional = false;
+                        bool     supportsOp1RegOptional = false;
+                        bool     supportsOp2RegOptional = false;
+                        bool     supportsOp3RegOptional = false;
+                        unsigned resultOpNum            = 0;
+                        LIR::Use use;
+                        GenTree* user = nullptr;
+
+                        if (BlockRange().TryGetUse(node, &use))
+                        {
+                            user = use.User();
+                        }
+                        resultOpNum = node->GetResultOpNumForFMA(user, op1, op2, op3);
+
+                        // Prioritize Containable op. Check if any one of the op is containable first.
+                        // Set op regOptional only if none of them is containable.
 
-                        if (IsContainableHWIntrinsicOp(node, op3, &supportsRegOptional))
+                        if (resultOpNum != 1 && IsContainableHWIntrinsicOp(node, op1, &supportsOp1RegOptional) &&
```

Comment:
As with https://github.com/dotnet/runtime/pull/58196/files#r751517074, I'd actually swap the check around containing `op1` with `op3`.

`op1` is the primary preference for `target` because it may be `copiesUpperBits` and it results in the least number of operand swaps; so it should be the least preferred for containment.

`op3` becomes the "most preferred" for containment because its the "least preferred" (last checked) for `resultOpNum` and its "naturally" in the containment slot (`op3` == `emitOp3`, if nothing is swapped).

Reply:
Do we swap only the ```MakeSrcContained``` part or both this and the ```SetRegOptional``` part?

Reply:
I think that would give us this:
```cpp
if ((resultOpNum != 3) && IsContainableHWIntrinsicOp(node, op3, &supportsOp3RegOptional))
{
    // result = (op1 * op2) + [op3]
    MakeSrcContained(node, op3);
}
else if (resultOpNum != 2 && IsContainableHWIntrinsicOp(node, op2, &supportsOp2RegOptional))
{
    // result = (op1 * [op2]) + op3
    MakeSrcContained(node, op2);
}
else if (resultOpNum != 1 && !HWIntrinsicInfo::CopiesUpperBits(intrinsicId))
{
    if (IsContainableHWIntrinsicOp(node, op1, &supportsOp1RegOptional))
    {
        // result = ([op1] * op2) + op3
        MakeSrcContained(node, op1);
    }
    else if (supportsOp1RegOptional)
    {
        op1->SetRegOptional();
    }
}
else if (supportsOp3RegOptional)
{
    assert(resultOpNum != 3);
    op3->SetRegOptional();
}
else if (supportsOp2RegOptional)
{
    assert(resultOpNum != 2);
    op2->SetRegOptional();
}
```

Reply:
I do think we want/need to be doing `if (supportsOp2RegOptional)` with `assert(resultOpNum != 2)` rather than the inverse as well.

Reply:
> Do we swap only the MakeSrcContained part or both this and the SetRegOptional part?

Its probably more consistent to swap both, but I don't think there's a strong need to do so for `RegOptional`. I've updated the suggested code above to account for keeping it consistent though.

The primary reason for asking for these to be swapped is to reduce the number of checks we need to do in the "normal" case. Due to how we set `resultOpNum` and the limits around `coppiesUpperBits`, this meant that `op1` is the "prime" candidate for the result register and so having that check first means it will also likely be false (unlikely to be taken) most of the time.

Once we get to `RegOptional` cases; we're mostly looking at it from a perspective of "if the register allocator has to spill something, can we tell it that some item is a better candidate for spilling than something else", in which case we're already in a "non ideal" situation and what we do probably doesn't have a lot of impact.